thirdparty/DiskANN/src/pq_flash_index.cpp

```diff
@@ -863,10 +848,171 @@ namespace diskann {
+        if (stats != nullptr) {
+          stats->n_4k++;
+          stats->n_ios++;
+        }
+      }
+      io_timer.reset();
+#ifdef USE_BING_INFRA
+      reader->read(frontier_read_reqs, ctx, true);  // async reader windows.
+#else
+      reader->read(frontier_read_reqs, ctx);    // synchronous IO linux
+#endif
+      if (stats != nullptr) {
+        stats->io_us += (double) io_timer.elapsed();
+      }
+
+      T *node_fp_coords_copy = data_buf;
+      for (const auto &req : frontier_read_reqs) {
+        const _u64 sector_offset = req.offset;
+        char      *sector_buf = reinterpret_cast<char *>(req.buf);
+        for (const auto cur_id : nodes_in_sectors_to_visit[sector_offset]) {
+          char *node_buf = get_offset_to_node(sector_buf, cur_id);
+          memcpy(node_fp_coords_copy, node_buf,
+                 disk_bytes_per_point);  // Do we really need memcpy here?
+          float dist =
+              dist_cmp(query, node_fp_coords_copy, (size_t) aligned_dim);
+          max_heap.Push(dist, cur_id);
+          if (feder != nullptr) {
+            feder->visit_info_.AddTopCandidateInfo(cur_id, dist);
+            feder->id_set_.insert(cur_id);
+          }
+        }
+      }
+
+      nodes_in_sectors_to_visit.clear();
+      frontier_read_reqs.clear();
+      sector_scratch_idx = 0;
+    }
+
+    for (_s64 i = k_search - 1; i >= 0; --i) {
+      if ((_u64) i >= max_heap.Size()) {
+        indices[i] = -1;
+        if (distances != nullptr) {
+          distances[i] = -1;
+        }
+        continue;
+      }
+      if (const auto op = max_heap.Pop()) {
+        const auto [dis, id] = op.value();
+        indices[i] = id;
+        if (distances != nullptr) {
+          distances[i] = dis;
+          if (metric == diskann::Metric::INNER_PRODUCT) {
+            distances[i] = 1.0 - distances[i] / 2.0;
+            if (max_base_norm != 0) {
+              distances[i] *= (max_base_norm * query_norm);
+            }
+          }
+        }
+      } else {
+        LOG_KNOWHERE_ERROR_ << "Size is incorrect";
+      }
+    }
+    if (stats != nullptr) {
+      stats->total_us = (double) query_timer.elapsed();
+    }
+    return;
+  }
+
+  template<typename T>
+  void PQFlashIndex<T>::cached_beam_search(
+      const T *query1, const _u64 k_search, const _u64 l_search, _s64 *indices,
+      float *distances, const _u64 beam_width, const bool use_reorder_data,
+      QueryStats *stats, const knowhere::feder::diskann::FederResultUniq &feder,
+      knowhere::BitsetView bitset_view) {
+    if (beam_width > MAX_N_SECTOR_READS)
+      throw ANNException("Beamwidth can not be higher than MAX_N_SECTOR_READS",
+                         -1, __FUNCSIG__, __FILE__, __LINE__);
+
+    ThreadData<T> data = this->thread_data.pop();
+    while (data.scratch.sector_scratch == nullptr) {
+      this->thread_data.wait_for_push_notify();
+      data = this->thread_data.pop();
+    }
+    auto query_norm_opt = init_thread_data(data, query1);
```

Comment:
Why we extract this out

Reply:
to make the function shorter, and also if it is easier to make a separate search function in the future